src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageprisonvisitsorchestration/client/VisitDetailsClient.kt

```diff
@@ -17,37 +19,90 @@ class VisitDetailsClient(
+        cancelledBy = userNames[lastVisit.cancelledBy] ?: lastVisit.cancelledBy,
+        updatedBy = userNames[lastVisit.updatedBy] ?: lastVisit.updatedBy,
+        createdDateAndTime = firstVisit.createdTimestamp,
+        updatedDateAndTime = getLastUpdatedVisitDateAndTime(visits),
+        cancelledDateAndTime = getCanceledVisitDateAndTime(lastVisit),
+        visit = lastVisit,
+      )
+    }
+
+    return null
+  }
+
+  private fun getLastUpdatedVisitDateAndTime(visits: List<VisitDto>): LocalDateTime? {
+    val lastVisitDto = visits.last()
+    if (visits.size > 1) {
+      return lastVisitDto?.let { it.createdTimestamp }
+    }
+    return null
+  }
+
+  private fun getCanceledVisitDateAndTime(lastVisit: VisitDto): LocalDateTime? {
+    return if (lastVisit.visitStatus == "CANCELED") {
+      lastVisit.modifiedTimestamp
+    } else {
+      null
     }
-    return visitDto
   }
 
-  private fun getUserDetails(actionedBy: String?): Mono<Optional<UserDetails>> {
-    actionedBy?.let {
-      // for past visits or some exceptional circumstances actionedBy will be NOT_KNOWN
-      if (actionedBy == NOT_KNOWN) {
-        return Mono.just(Optional.of(UserDetails(actionedBy, NOT_KNOWN)))
-      } else {
-        return hmppsAuthClient.getUserDetails(actionedBy)
+  private fun getUserDetails(visitDto: VisitDto): Map<String, String> {
+    val monoCallsList = createUserMonoCalls(visitDto)
+    return executeMonoCalls(monoCallsList)
+  }
+
+  private fun executeMonoCalls(monoCallsList: List<Mono<UserDetailsDto>>): Map<String, String> {
+    var results: MutableMap<String, String> = mutableMapOf()
+    if (monoCallsList.size == 1) {
+      val userDetails = monoCallsList[0].block(apiTimeout)
+      userDetails.fullName?.let { results[userDetails.username] = userDetails.fullName }
+    }
+
+    if (monoCallsList.size > 1) {
+      var zipResults: MutableList<UserDetailsDto> = mutableListOf()
+      if (monoCallsList.size == 2) {
+        val userDetailsTuples = Mono.zip(monoCallsList[0], monoCallsList[1]).block(apiTimeout)
+        zipResults = userDetailsTuples.toList() as MutableList<UserDetailsDto>
+      }
+      if (monoCallsList.size == 3) {
+        val userDetailsTuples = Mono.zip(monoCallsList[0], monoCallsList[1], monoCallsList[2]).block(apiTimeout)
+        zipResults = userDetailsTuples.toList() as MutableList<UserDetailsDto>
       }
+      zipResults?.forEach { userDetails -> userDetails.fullName?.let { results[userDetails.username] = userDetails.fullName } }
     }
 
-    return Mono.just(Optional.empty())
+    return results.toMap()
   }
 
-  private fun getFullName(userDetails: Optional<UserDetails>?): String? {
-    return if (userDetails?.isEmpty == true) null else userDetails?.get()?.name
+  private fun createUserMonoCalls(
+    visitDto: VisitDto,
+  ): List<Mono<UserDetailsDto>> {
+    val userNames = mutableSetOf(visitDto.createdBy, visitDto.updatedBy, visitDto.cancelledBy)
+    userNames.remove(null)
```

Comment:
minor comment - we could combine this on Line 93. Trivial, can be ignored.

Reply:
`public abstract fun remove(
    element: E
): Boolean`

remove is not builder pattern will see if I can find something else.

Reply:
Ignore. I thought it might be.